### PR TITLE
New ctl verbs

### PR DESCRIPTION
**Pull request by sam-leonard-ct**
Adds verbs for both Journalctl and coredumpctl.

### CODE REVIEW

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
We should use --root if we're doing a directory image and we should show an error if the output is neither disk nor directory

**Comment by sam-leonard-ct** [comment]
I've pushed a fix, I haven't squashed it into the last one but am obviously very happy to do so if you would prefer.

File: mkosi/config.py [hunk 1]
```diff
@@ -56,15 +56,20 @@ class Verb(StrEnum):
...
 
     def needs_build(self) -> bool:
-        return self in (Verb.build, Verb.shell, Verb.boot, Verb.qemu, Verb.serve)
+        return self in (Verb.build, Verb.shell, Verb.boot, Verb.qemu,
+                        Verb.serve, Verb.journalctl, Verb.coredumpctl)
 
     def needs_root(self) -> bool:
-        return self in (Verb.shell, Verb.boot)
+        return self in (Verb.shell, Verb.boot, Verb.journalctl,
+                        Verb.coredumpctl)
```
Comment (behrmann) [inline mkosi/config.py:72]: No need to break the lines here, all of them are short enough to fit on a single line within our current line limit.

**Review by behrmann — APPROVED** [review]
I think it would make sense to squash the whole series into a single commit.

File: mkosi/config.py (hunk 1 above)
Comment (sam-leonard-ct) [reply mkosi/config.py:72]: Okay I will fix these and squash it, this was just where my editor broke it and I assumed it had found some config file somewhere to tell it to do that 🤷

Comment (sam-leonard-ct) [reply mkosi/config.py]: This is fixed now and the commits have been squashed.

File: mkosi/__init__.py [hunk 2]
```diff
@@ -2292,6 +2292,31 @@ def run_shell(args: MkosiArgs, config: MkosiConfig) -> None:
...
+    if (tool_path := find_binary(tool)) is None:
+        die(f"Failed to find {tool}")
+
+    image_arg_name = "root" if config.output_format == OutputFormat.directory else "image"
+    fname = config.output_dir_or_cwd() / config.output
+    cmdline: list[PathString] = [
+        tool_path,
+        f"--{image_arg_name}={fname}",
+        *args.cmdline
+    ]
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:2308]: This doesn't need its own variable, please inline it in the run() call

File: mkosi/__init__.py [hunk 3]
```diff
@@ -2292,6 +2292,31 @@ def run_shell(args: MkosiArgs, config: MkosiConfig) -> None:
...
 
+def run_ctl_with_name(tool: str, args: MkosiArgs, config: MkosiConfig) -> None:
+    if config.output_format not in (OutputFormat.disk, OutputFormat.directory):
+        die(f"{config.output_format} images cannot be inspected with {tool}")
+
+    if (tool_path := find_binary(tool)) is None:
+        die(f"Failed to find {tool}")
+
+    image_arg_name = "root" if config.output_format == OutputFormat.directory else "image"
+    fname = config.output_dir_or_cwd() / config.output
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:2303]: This variable doesn't really have much use so I'd inline it

File: mkosi/__init__.py [hunk 4]
```diff
@@ -2660,3 +2685,9 @@ def run_verb(args: MkosiArgs, images: Sequence[MkosiConfig]) -> None:
 
             if args.verb == Verb.serve:
                 run_serve(last)
+
+            if args.verb == Verb.journalctl:
+                return run_journalctl(args, last)
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:2690]: Let's drop the return here and below since we don't do it for run_serve() either

File: mkosi/config.py [hunk 5]
```diff
@@ -56,15 +56,17 @@ class Verb(StrEnum):
...
-        return self in (Verb.build, Verb.shell, Verb.boot, Verb.qemu, Verb.ssh)
+        return self in (Verb.build, Verb.shell, Verb.boot, Verb.qemu, Verb.ssh, Verb.journalctl, Verb.coredumpctl)
 
     def needs_build(self) -> bool:
-        return self in (Verb.build, Verb.shell, Verb.boot, Verb.qemu, Verb.serve)
+        return self in (Verb.build, Verb.shell, Verb.boot, Verb.qemu, Verb.serve, Verb.journalctl, Verb.coredumpctl)
 
     def needs_root(self) -> bool:
-        return self in (Verb.shell, Verb.boot)
+        return self in (Verb.shell, Verb.boot, Verb.journalctl, Verb.coredumpctl)
```
Comment (DaanDeMeyer) [inline mkosi/config.py:69]: Do we need root if we're invoking this on a directory image?

File: mkosi/__init__.py [hunk 6]
```diff
@@ -2292,6 +2292,31 @@ def run_shell(args: MkosiArgs, config: MkosiConfig) -> None:
         run(cmdline, stdin=sys.stdin, stdout=sys.stdout, env=os.environ, log=False)
 
 
+def run_ctl_with_name(tool: str, args: MkosiArgs, config: MkosiConfig) -> None:
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:2295]: I'd name this run_systemd_tool()

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/__init__.py (hunk 2 above)
Comment (behrmann) [reply mkosi/__init__.py:2308]: It's funny. I was something similar yesterday, but when testing whether the type annotation could be omitted mypy wasn't smart enough to not make this `list[Object]`, but it works when inlining. Weird edge case.

Comment (sam-leonard-ct) [reply mkosi/__init__.py:2308]: behrmann I also had that exact issue yesterday 😂 and when I had some slightly more complicated logic for the cmdline that was all that fixed it. I will inline it now :)

File: mkosi/config.py (hunk 5 above)
Comment (sam-leonard-ct) [reply mkosi/config.py:69]: I'm not sure, I'm just going for lunch and will test in an hour or so.

Comment (sam-leonard-ct) [reply mkosi/config.py:69]: Nope it doesn't, I have added a check as early as possible for it here: https://github.com/CodethinkLabs/mkosi/blob/bb63b8d74e4379f526fe0093de9370ffed270dc1/mkosi/__init__.py#L2591

**Pull request merged**